Improve automatic language boxes

ptersilie opened the pull request:
**Requires merging #243 first.**

This PR adds various improvements and additions to automatic language boxes:
- e5653c8: Allows language boxes to expand if tokens following the box can be moved inside the box. This commit also improves loose/fixed boxes by fixating automatically inserted (loose) language boxes when the user manually moves the cursor out of them.
- e4f5bc3: Extends e5653c8. If there are multiple expansion possibilities, they are shown via the lightbulb icon on the left.
- 4df519f: Added long overdue option to remove language boxes and interpret their content in the outer language.
- 8f39a27: Use error recovery during the automatic removal check.
- 78c5943: Allows language boxes to shrink again if they contain a parse error.
- 830798d: Fixes e5653c8 to only fixate language boxes if they were manually chosen by the user.
- 17a426b: Fix: Makes sure that automatically inserted language boxes are reparsed in a single step (previously we would reparse after deletion of the old nodes and then again after inserting the language box).
- 437fd5f: Improves e5653c8. Only expand language boxes if this doesn't introducing errors immediately after the box.

ptersilie committed:
Show multiple expand options in the editor

If there are multiple ways in which a language box can be expanded, show
all options via the familiar lightbulb icon.

ptersilie committed:
Add menu option to remove language boxes

This commit allows users to remove a selected language box via the
editor's menu, which moves its contents into the outside box.

After removing a language box, also make sure to relex tokens just
before the box as they may need to be merged with the tokens from the
box.

ptersilie committed:
Check for removable language boxes during parsing

Since recognisers have no error recovery, isolated changes are skipped
during these checks, thus ignoring changes which could trigger the
removal of a box. Moving the removal check directly into the parser,
gives us access to error recovery which solves this problem.

ptersilie committed:
Automatically shrink invalid language boxes

If a language box contains an error check if it contains a valid subset
and reduce the extensions of the language box to that subset, moving the
remaining code outside of the box.

ptersilie committed:
Limit reparsing during automatic insertions

Only reparse once a language box has been inserted and the code it
replaces has been moved inside the box. On other words: Don't reparse
immediately after the existing code has been removed, but wait until it
has been moved into the new language box.

ptersilie commented:
Addressed comments.

ptersilie commented:
Added a fix for the automatic language box expansion, which wasn't working properly for some examples.

ltratt commented on lib/eco/treemanager.py:
```diff
@@ -2191,16 +2191,17 @@ def lbox_expand_test(self, lbox):
+            if r2.parse_after(e.next_term):
                 filtered.append(e)
-            start = e
+            r2.state = tmp
```
Why are we (in effect) doing `r2.state = list(r2.state)`? What's the `list` call for?

ptersilie replied:
`state` is the (admittedly poorly named) parsing stack of the recogniser. After initialising the recogniser with `preparse`, we then want to try and check if each element in `possible_end` can be parsed in that state, so we have to reset the parse stack back after every iteration.

ltratt replied:
Yes, but it's the `list` bit that I'm confused by.

ptersilie replied:
It just creates a copy of the parse stack which we can then later apply to r2, to reset the parse stack back to what it was before.

ptersilie replied:
I can also use `r2.state[:]` if you prefer that.

ltratt replied:
Oh, I see, you're doing a shallow copy! Ah. Yes, `[:]` will make this clearer. I thought you were converting from a non-list to a list!

ptersilie replied:
Fixed in the latest commit.

ltratt commented:
Tests failed?

ptersilie commented:
Weird, looks fine on travis. I'll just restart.

ptersilie commented:
Nevermind, found it now. Will fix.

ptersilie commented:
Okay, this is taking a bit longer than I expected. I'll let you know when this is fixed.

ltratt commented:
OK.

ptersilie committed:
Various fixes for the history based heuristic

1) Since the heuristic is based on the previous parse tree we also need
to use the previous version during the `preparse` method.

2) While searching for the left neighbor of the error subtree, skip
empty nonterminals as we later need to use the most-right terminal
inside `left` to put the parser into the correct state.

ptersilie committed:
Temporarily disable automatic language boxes

When inserting, removing, or resizing a language box, temporarily
disable automatic language box detection to avoid chain reactions, which
can lead to problems.

ptersilie commented:
This should fix things. I also had to rebase as the error would only appear when merged into master.

ltratt commented on lib/eco/autolboxdetector.py:
```diff
@@ -97,16 +97,22 @@ def detect_lbox_h2(self, errornode):
                             term = term.next_term
+                        element = self.op.syntaxtable.lookup(term.state, lbox)
+                        if type(element) not in [Reduce, Shift]:
+                            # can we still shift lbox?
```
What does the comment mean?

ptersilie replied:
We have already checked that the lbox can be shifted after the nonterminal `left`. This is a sanity check, that tests if lbox can still be shifted after `term` which is the most-right terminal in `left`. This should almost always be true, so this double checks that this is really the case. I believe there can be cases, where the terminal is an error node, in which case this second check will fail, so that we don't insert a box which then turns out to be invalid when reparsed.

ltratt replied:
Do we have a test case for this? If not, can we maybe expand the comment along the lines of what you've just written.

ptersilie replied:
I think I need to add some tests for this and the other features, as I just found another bug with the new stuff.

ptersilie replied:
Although in this case, I don't currently have a test, though I might be able to come up with one.

ltratt replied:
Let's give it a go.

ptersilie commented:
I've just added a couple of tests for the new improvements, and also clarified the comment for which I couldn't find a test.

ltratt commented on lib/eco/test/test_eco.py:
```diff
@@ -5247,3 +5247,44 @@ def test_java_lua_dont_remove_explicit_lboxes(self):
+        for c in p:
+            treemanager.key_normal(c)
+        assert len(treemanager.parsers) == 1
+        assert parser.last_status == True
```
You don't need `== True` yet.

ltratt replied:
s/yet/here/

ltratt commented on lib/eco/test/test_eco.py:
```diff
@@ -5247,3 +5247,44 @@ def test_java_lua_dont_remove_explicit_lboxes(self):
+        treemanager.key_backspace()
+
+        assert len(treemanager.parsers) == 2 # box now has been expanded
+        assert parser.last_status == True
```
Ditto.

ltratt commented:
Where are with this one?

ptersilie commented:
I'm almost done going through the logs and fixing bugs / adding features when I encounter problems.

ptersilie commented:
Alright, I think we can get this squashed and merged now.

ltratt commented:
OK I checked the new commits and they look OK. Please squash.

ptersilie committed:
Permanently fix loose language boxes

When the user manually moves the cursor outside of an automatically
inserted (loose) language box, that box is permanently fixed, which means it
will not be automatically removed or resized again. We should later add
an option for the user to manually re-mark language boxes as loose.

ptersilie committed:
Automatically expand existing language boxes

This commit enables the editor to automatically expand loose language
boxes, by moving code following the language box to the inside.
Currently, a language box always tries to contain the maximum amount of
valid code, although an alternative option would be to only expand a box
if the outside code would otherwise be invalid.

ptersilie committed:
Added tests for automatic language box shrinking

ptersilie committed:
Added tests for automatic language box expansion

ptersilie commented:
Squashed!
